hwdec_surfacetexture for Android/MediaCodec

tmm1 opened the pull request:
Implements a rw_hwdec driver for IMGFMT_MEDIACODEC on Android, as suggested by @wang-bin in https://github.com/mpv-player/mpv/pull/4580#issuecomment-358844146

This make `--vo=gpu --hwdec=mediacodec` work (as opposed to `--hwdec=mediacodec-copy`), by downloading the video frame from the MediaCodec decoder into an RGB GLES texture (via GL_TEXTURE_EXTERNAL_OES).

SurfaceTexture has a weird transform matrix requirement that is described in https://developer.android.com/reference/android/graphics/SurfaceTexture.html:

>When sampling from the texture one should first transform the texture coordinates using the matrix queried via getTransformMatrix(float[]). The transform matrix may change each time updateTexImage() is called, so it should be re-queried each time the texture image is updated. This matrix transforms traditional 2D OpenGL ES texture coordinate column vectors of the form (s, t, 0, 1) where s and t are on the inclusive interval [0, 1] to the proper sampling location in the streamed texture. This transform compensates for any properties of the image stream source that cause it to appear different from a traditional OpenGL ES texture. For example, sampling from the bottom left corner of the image can be accomplished by transforming the column vector (0, 0, 0, 1) using the queried matrix, while sampling from the top right corner of the image can be done by transforming (1, 1, 0, 1).

The transform matrix has to be passed into the vertex shader, which I got working but it's hacked in pretty badly. I could use some advice on how to integrate it more cleanly, so it doesn't affect other gpu renderers.

There are some remaining issues:

- [x] transform matrix multiplication needs to be conditional in the vertex shader
- [x] video is flipped for some reason
- [x] playback stalls after pause/play
- [x] integrate `setOnFrameAvailableListener`
- [x] on some devices, video stalls after first few frames with "could not consume packet" errors
- [x] visual artifacts on some devices
- [ ] need `VO_CAP_NORETAIN` equivalent on vo_gpu
- [ ] pthread destructor is SEGVing on @sfan5's device

cc @haasn @sfan5 @xyzz @jeeb @wm4 @wang-bin

tmm1 commented on video/out/opengl/egl_helpers.c:
```diff
@@ -193,7 +193,11 @@ bool mpegl_create_context_cb(struct ra_ctx *ctx, EGLDisplay display,
 
+#if HAVE_ANDROID
+    int es[] = {3, 2, 0}; // preference order
+#else
```
Some Android TV devices offer desktop opengl, but that doesn't work with the GLES GL_TEXTURE_EXTERNAL_OES extension. I had to switch the preference order here to make the hwdec work.

tmm1 replied:
I think an alternative might be --opengl-es=yes

tmm1 commented:
> video is flipped for some reason

Was able to fix this by changing the transform matrix provided by SurfaceTexture, h/t https://stackoverflow.com/questions/30595493#comment53190986_30635539

> on some devices, video stalls after first few frames with "could not consume packet" errors

Adding `VO_CAP_NORETAIN` to vo_gpu fixes this, so the problem is too many mediacodec frames are being buffered again. I think the existing checks for NORETAIN can be rewritten to check for IMGFMT_MEDIACODEC instead.

tmm1 commented on video/out/opengl/hwdec_surfacetexture.c:
```diff
@@ -0,0 +1,343 @@
+    AVMediaCodecBuffer *buffer = (AVMediaCodecBuffer *)mapper->src->planes[3];
+    av_mediacodec_release_buffer(buffer, 1);
+
+    JNI_CALL_VOID(o->texture, SurfaceTexture.updateTexImage);
```
To map the mediacodec frame to a texture, I'm calling `MediaCodec#releaseOutputBuffer(render=1)` followed immediately by `SurfaceTexutre#updateTexImage`.

This seems to work fine on fast devices, but on slower ones I'm noticing video artifacts during playback. I think that must be because there's some latency before the texture becomes available, and we're supposed to use `SurfaceTexture#setOnFrameAvailableListener` before calling `updateTexImage`. That requires a java listener though, and happens asynchronously, so I'm not sure how best to integrate it into this JNI wrapper.

tmm1 replied:
> That requires a java listener though, and happens asynchronously, so I'm not sure how best to integrate it into this JNI wrapper.

Seems the only way to do this is to require the libmpv user to include a java file in their project. Here's an example:

- user would include [OnFrameAvailableListener.java](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android/java/OnFrameAvailableListener.java) in their java project
- libmpv would [fetch that class at runtime](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android/java/OnFrameAvailableListener.java) and [bind it to a C callback](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L50)
- the C callback would [signal a condition](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L456-L457) and the hwdec_surfacetexture mapper would [wait on the condition after calling releaseOutputBuffer](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L394-L411)

haasn commented:
Rather than hacking the transform into the texture and multiplying from inside the vertex shader, you can most likely update the texture coordinates on the vertices directly. The easiest way of doing this is multiplying the matrix into the image transform in `pass_get_images` (probably by checking for `IMGFMT_MEDIACODEC` I'd guess?).

tmm1 commented:
> Rather than hacking the transform into the texture and multiplying from inside the vertex shader, you can most likely update the texture coordinates on the vertices directly. The easiest way of doing this is multiplying the matrix into the image transform in pass_get_images (probably by checking for IMGFMT_MEDIACODEC I'd guess?).

Thanks. I noticed the `gl_transform` calls and was wondering if I could hook into that somehow.

The transform matrix from mediacodec is per-frame, and by putting it in `ra_tex_params` I was able to access it easily from `pass_prepare_src_tex`.

I'm trying to figure out where I would put it instead so that I could access it from `pass_get_images`. There's a lot of indirection and structs involved in this pipeline, so if you can point me in the right direction that would be helpful.

haasn commented:
You can probably put it into `struct ra_hwdec_mapper`, next to the `struct ra_tex *tex[4]; bool vdpau_fields;`. Just add a new `gl_transform` here and multiply it into the plane transforms if hwdec is being used.

haasn commented:
I guess it depends on whether those fields are expected to change per frame or not. Worst case scenario, you can always add a new pointer `out_transform` to `ra_hwdec_mapper_map` or whatever. @wm4 probably has a better opinion here. At any rate, I don't think it belongs in `ra_tex`

wm4 commented:
The `tex` and `vdpau_fields` fields in `struct ra_hwdec_mapper` can change on every map/unmap, and new fields could follow the same rules.

tmm1 commented:
Thanks for the advice!

I added a `transform` property to `ra_hwdec_mapper` and wired it up in `pass_get_images`. The vertex shader hack is gone now.

haasn commented:
Is the check for IMGFMT_MEDIACODEC even needed now? Why not just always apply the transform if a hwdec is present? (Although we would need to find a good place to have it default to `identity_trans` for the other hwdecs)

Also, won't that check segfault when not using hwdec?

wm4 commented:
>Also, won't that check segfault when not using hwdec?

Yes, it will. I'd prefer if this were done in `pass_upload_image`, where it inits `vimg->planes` from the mapper.

tmm1 commented:
I refactored the JNI bindings in this PR using the same helpers as in #5589.

Also added the `setOnFrameAvailableListener` hook, though on many devices it seems to work fine without. So I made it optional- if you don't specify a java listener class, it will warn you and tell you how to set it up but still continues anyway.

I tried this on ~5 different Android TV devices and it works well on all except one.

> on some devices, video stalls after first few frames with "could not consume packet" errors
> visual artifacts on some devices

I still see these issues on one specific device (gen3 FireTV), but it appears to be because the hardware/drivers are broken. I tried another android media player which has a texture mapping option, and it fails in a similar way. Strange, an older model (gen2 FireTV stick) works just fine.

cc @sfan5 @jeeb @xyzz Could I get you to try this rendering mode on some of your devices?

tmm1 commented:
@sfan5 said this worked on his device too.

This PR is ready for review/merge.

wang-bin commented on video/out/opengl/hwdec_surfacetexture.c:
```diff
@@ -0,0 +1,432 @@
+    (*env)->GetFloatArrayRegion(env, p->matrix, 0, 16, p->transform);
+    if (MP_JNI_EXCEPTION_LOG(mapper->owner)) return -1;
+
+    mapper->transform = (struct gl_transform){
```
I doubt the transform is wrong. let M be the matrix from getTransformMatrix, F be the flip matrix, M maps sampler point (x,y,..) to texture point F*(x,y, ..), so the result matrix should be M*F^-1=M*F. But i don't know how mpv uses tranform in shader.

Do you have an prebuilt apk for testing?

wang-bin replied:
mediacodec can not be enabled in this apk

tmm1 commented:
> on some devices, video stalls after first few frames with "could not consume packet" errors
> visual artifacts on some devices

I was able to figure this out. The artifacts were caused by the "could not consume packet", which was mpv saying that it was dropping an input packet. That's why the output was also corrupted.

The fix was ffmpeg/ffmpeg@3172b31223e1f0a64233fa5b355e11d7a00e477b

> @sfan5 said this worked on his device too.

I misunderstood earlier, and it's actually _not_ working on his device. There's some sort of color skew issue happening on Snapdragon Adreno GPUs, seen on both a gen1 FireTV as well as Samsung Galaxy tablet.

Expected:
![image](https://user-images.githubusercontent.com/2567/37370413-5f90f22a-26c9-11e8-8ccd-87be4e5c42b1.png)

Actual:
![image](https://user-images.githubusercontent.com/2567/37370436-78f4dd76-26c9-11e8-922d-855c6d72749c.png)

I'm struggling to figure out what could be causing this (and only on this specific GPU), so any hunches would be appreciated.

tmm1 commented:
> I'm struggling to figure out what could be causing this (and only on this specific GPU), so any hunches would be appreciated.

@sfan5 did some sleuthing with GIMP, and by mapping G -> Y, B -> U, R -> V he was able to fix the colors on the generated image.

According to https://stackoverflow.com/a/46257292/332798, SurfaceTexture is supposed to convert the GL_TEXTURE_EXTERNAL_OES from its native format to RGB transparently when accessed via the fragment shader. However, it seems that on Adreno GPUs this is not happening? /cc @mstorsjo have you heard of this before?

mstorsjo commented:
Nope, haven't heard about this issue. And after all, Android has got a CTS test since 4.3 that mandates this case to work properly, otherwise devices aren't allowed to be shipped claiming that they're Android compatible.

So something differs between how it's used here compared to what's tested and most probably works in those tests. The tests iirc render the texture to a pbuffer surface and then run glReadPixels() on it to check the RGB values.

If you add a glReadPixels() in this setup and check the values, do they match the expected RGB values, or is it the raw YUV values here as well? If glReadPixels() gets the right values, something is amiss with the output of the rendering context, while it just otherwise is with the rendering context as a whole or how the input texture is hooked up.

tmm1 commented:
Thanks @mstorsjo, that's really helpful and it saves me from chasing a red herring.

I have a theory that this might be related to the fact that I'm using `texture()` instead of `texture2D()`, and the former is not performing the automatic conversion on these GPUs.

tmm1 commented:
> I have a theory that this might be related to the fact that I'm using texture() instead of texture2D(), and the former is not performing the automatic conversion on these GPUs.

`texture()` is just the gles3 version of `texture2D()`, so that theory doesn't make any sense.

Here's what the generated fragment shader looks like:

```
[  1] #version 300 es
[  2] #extension GL_OES_EGL_image_external : enable
[  3] precision mediump float;
[  4] precision mediump sampler2D;
[  5] precision mediump sampler3D;
[  6] #define tex1D texture
[  7] #define tex3D texture
[  8] #define LUT_POS(x, lut_size) mix(0.5 / (lut_size), 1.0 - 0.5 / (lut_size), (x))
[  9] out vec4 out_color;
[ 10] in vec2 texcoord0;
[ 11] uniform mat3 colormatrix;
[ 12] uniform vec3 colormatrix_c;
[ 13] uniform vec3 src_luma;
[ 14] uniform vec3 dst_luma;
[ 15] uniform samplerExternalOES texture0;
[ 16] uniform vec2 texture_size0;
[ 17] uniform mat2 texture_rot0;
[ 18] uniform vec2 texture_off0;
[ 19] uniform vec2 pixel_size0;
[ 20] void main() {
[ 21] vec4 color = vec4(0.0, 0.0, 0.0, 1.0);
[ 22] color.rgb = 1.000000 * vec4(texture(texture0, texcoord0)).rgb;
[ 23] color = color.rgbr;
[ 24] color.rgb = mat3(colormatrix) * color.rgb + colormatrix_c;
[ 25] color.a = 1.0;
[ 26] // color mapping
[ 27] color.rgb *= vec3(1.000000);
[ 28] color.rgb *= vec3(1.000000);
[ 29] out_color = color;
[ 30] }
```

It's doing some funky stuff with the color after the `texture()` call, so that might be where things are going awry. Not sure why it would only affect certain GPUs though.

mstorsjo commented:
I don't see anything strange about the colormatrix stuff, I doubt that's where it goes wrong. That matrix could however be used to convert rgb to yuv, but since this is an unintended and undocumented condition, I wouldn't rely on it.

One potential suspect is the fact that SurfaceTexture is initialized with an invalid texture and then detached. You could try to construct it and start the mediacodec decoder while the texture is attached. It's concievable that the yuv->rgb conversion is enabled at that point, and if the SurfaceTexture isn't attached to the EGL context and GL texture object at this point, this detail gets lost. (I'm well aware that this can require some tricky restructuring of the code.)

mstorsjo commented:
I would advice against assuming that the output is yuv in this case. Before this, I had never heard that it was possible to get yuv in this codepath at all, so it's probably a very obscure corner case you're hitting. I'd rather suggest figuring out what is so dramatically different in your use of mediacodec+surfacetexture.

Or on the other hand, if doing it this way, be prepared that it might start working (and thus get broken by this workaround) on adreno in older or newer firmware versions.

tmm1 commented:
> I don't see anything strange about the colormatrix stuff, I doubt that's where it goes wrong. That matrix could however be used to convert rgb to yuv, but since this is an unintended and undocumented condition, I wouldn't rely on it.

You're right. I removed all that code from the shader, and the output image was still wrong. So the mpv shader code was not to blame.

I ended up hacking in a check for the Adreno GPU which forces the hwdec to interpret the texture as VYU, and then uses the colormatrix to convert back to RGB (@sfan5 just pushed it to this PR for me).. It's still unclear why this is required, but it fixes the color issue on both our devices.

> One potential suspect is the fact that SurfaceTexture is initialized with an invalid texture and then detached. You could try to construct it and start the mediacodec decoder while the texture is attached. It's concievable that the yuv->rgb conversion is enabled at that point, and if the SurfaceTexture isn't attached to the EGL context and GL texture object at this point, this detail gets lost. (I'm well aware that this can require some tricky restructuring of the code.)

Interesting.. this approach was initially suggested by @wang-bin in https://github.com/mpv-player/mpv/pull/4580#issuecomment-315496087. I think you may be right that this is what's causing the issue, because I'm struggling to see what else is different between my code and all the examples I can find.

Restructuring the code to initialize the SurfaceTexture (and thus MediaCodec) until the opengl context is created is a bit tricky. I'll see if I can try that, if only just to verify that it is indeed what is causing the discrepancy.

tmm1 commented:
> Restructuring the code to initialize the SurfaceTexture (and thus MediaCodec) until the opengl context is created is a bit tricky. I'll see if I can try that, if only just to verify that it is indeed what is causing the discrepancy.

Gave this a shot in https://gist.github.com/tmm1/b25f9373344f77a3f3e996d332625f7a. It was easier than I expected, but unfortunately it made no difference.

mstorsjo commented:
I've got a phone with an Adreno chip, and I could try to see if I manage to reproduce it (either with the mpv apk file, or with a generic mediacodec example modified to use SurfaceTexture in this way). I tested https://0x0.st/sacr.apk from yesterday (downloaded a moment ago), but it seems like it doesn't seem to pick up using mediacodec here:

```
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v] Codec list:
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v]     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v]     h264_mediacodec (h264) - H.264 Android MediaCodec decoder
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v] Opening decoder h264
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v] Looking at hwdec h264_mediacodec-mediacodec...
03-14 21:54:24.123 31885  2201 V mpv     : [vo/gpu:v] Loading hwdec driver 'cuda-nvdec'
03-14 21:54:24.123 31885  2201 V mpv     : [vo/gpu/cuda-nvdec:v] Failed to load CUDA symbols
03-14 21:54:24.123 31885  2201 V mpv     : [vo/gpu:v] Loading failed.
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v] Could not create device.
03-14 21:54:24.123 31885  2201 V mpv     : [vd:v] Using software decoding.
03-14 21:54:24.124 31885  2201 V mpv     : [vd:v] Detected 5 logical cores.
03-14 21:54:24.124 31885  2201 V mpv     : [vd:v] Requesting 6 threads for decoding.
03-14 21:54:24.128 31885  2201 V mpv     : [vd:v] Selected codec: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
```

tmm1 commented:
Here's an updated apk that I just built. It has the adreno hack built in, and debug logging on so you can see the shader source: https://www.dropbox.com/s/gmyaaqr7g7p28r0/mpv-android-surfacetexture.apk?dl=1

Happy to build another without the hack if you need it.

mstorsjo commented:
> Here's an updated apk that I just built. It has the adreno hack built in, and debug logging on so you can see the shader source: https://www.dropbox.com/s/gmyaaqr7g7p28r0/mpv-android-surfacetexture.apk?dl=1

Thanks - I tested this and now I see the hw decoder being enabled and used, and I'm getting incorrect colors with it.

```
03-14 23:06:53.205 13435 13942 V mpv     : [vd:v] Container reported FPS: 24.000000
03-14 23:06:53.208 13435 13942 V mpv     : [vd:v] Codec list:
03-14 23:06:53.208 13435 13942 V mpv     : [vd:v]     h264 - H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v]     h264_v4l2m2m (h264) - V4L2 mem2mem H.264 decoder wrapper
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v]     h264_mediacodec (h264) - H.264 Android MediaCodec decoder
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v] Opening decoder h264
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v] Looking at hwdec h264_mediacodec-mediacodec...
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v] Trying hardware decoding via h264_mediacodec-mediacodec.
03-14 23:06:53.209 13435 13942 V mpv     : [vd:v] Using underlying hw-decoder 'h264_mediacodec'
03-14 23:06:53.223 13435 13942 V mpv     : [vd:v] Pixel formats supported by decoder: mediacodec
03-14 23:06:53.223 13435 13942 V mpv     : [vd:v] Codec profile: unknown (0xffffff9d)
03-14 23:06:53.223 13435 13942 V mpv     : [vd:v] Requesting pixfmt 'mediacodec' from decoder.
```

I don't really see in the log whether the workaround triggered (and thus my device is behaving correctly and the workaround breaks it), or if the workaround failed to kick in (and my device is consistent with other adreno devices).

My phone is a Xperia Z5 C (Snapdragon 810) running Android 7.1.1, fwiw.

tmm1 commented:
> I don't really see in the log whether the workaround triggered (and thus my device is behaving correctly and the workaround breaks it), or if the workaround failed to kick in (and my device is consistent with other adreno devices).

Thanks for testing. It sounds like your device is returning RGB as expected.

If you look at the fragment shader source, you'll see either `color.rgbr` or `color.gbrr` (adreno hack).

>[vo/gpu/opengl:debug] [ 22] color = color.rgbr;

mstorsjo commented:
> If you look at the fragment shader source, you'll see either color.rgbr or color.gbrr (adreno hack).

I'm seeing `[ 23] color = color.gbrr;`, so it seems like the hack is invoked, and consequently breaks things. I guess that means I can't unfortunately help with hands-on debugging of the issue. :-(

mstorsjo commented:
(Also unrelatedly - I don't see any files in the main browser interface, I'm only seeing "mpv :: /" and a listing containing "..". I'm able to play a test file by manually entering e.g. `/sdcard/test.mp4` though.)

tmm1 commented:
> I'm seeing [ 23] color = color.gbrr;, so it seems like the hack is invoked, and consequently breaks things. I guess that means I can't unfortunately help with hands-on debugging of the issue. :-(

No problem, you've been plenty helpful already.

I've updated this PR to remove the `detachGLContext`, added a log message when the Adreno hack is enabled, and also limited it to just Adreno 3xx GPUs.

Here's a new apk which should output the correct colors on your device (by skipping the hack): https://www.dropbox.com/s/4lucki25aviqri5/mpv-android-surfacetexture-adreno3xx.apk?dl=1

mstorsjo commented:
> I've updated this PR to remove the detachGLContext, added a log message when the Adreno hack is enabled, and also limited it to just Adreno 3xx GPUs.
> 
> Here's a new apk which should output the correct colors on your device (by skipping the hack): https://www.dropbox.com/s/4lucki25aviqri5/mpv-android-surfacetexture-adreno3xx.apk?dl=1

This APK seems to be identical to the previous one, so I'm still getting wrong colors (but no log message telling me about it).

For some reason, the directory listing suddenly started working though...

tmm1 commented:
> This APK seems to be identical to the previous one, so I'm still getting wrong colors (but no log message telling me about it).

Sorry I must have copied the wrong file (had the same sha1 as the old one). I replaced it, so try downloading it again.

> For some reason, the directory listing suddenly started working though...

I noticed this too (directory listing worked on second launch). I don't know much about the mpv-android UI code as I mostly work on libmpv itself.

mstorsjo commented:
> Sorry I must have copied the wrong file (had the same sha1 as the old one). I replaced it, so try downloading it again.

The old link at https://www.dropbox.com/s/4lucki25aviqri5/mpv-android-surfacetexture-adreno3xx.apk?dl=1 now gives 404, so apparently dropbox wants a new url if the file changed?

tmm1 commented:
Hm, I guess it's a content based URI.

https://www.dropbox.com/s/j74oivvf8yg10x0/mpv-android-surfacetexture-adreno3xx.apk?dl=1

mstorsjo commented:
> https://www.dropbox.com/s/j74oivvf8yg10x0/mpv-android-surfacetexture-adreno3xx.apk?dl=1

Thanks - this one does seem to work as it should on my device.

shinchiro commented:
Would like to test this on my amlogic s905 android tv. Does this apk should be good?
> https://www.dropbox.com/s/j74oivvf8yg10x0/mpv-android-surfacetexture-adreno3xx.apk?dl=1

tmm1 commented:
Yes that apk should work. I tried it on an AMLogic based MiBox3

shinchiro commented:
So far, I tested with Mali-400 MP and 2 device with Adreno 306. Tested with 8bit h264 video and all of them can use mediacodec fine. But one of the adreno 306 (vivo y25) show lots of banding compared to other devices. Is it related?

vivo y25's logcat: https://0x0.st/sajO.txt

sfan5 commented:
> show lots of banding compared to other devices

like this? mpv-android/mpv-android#112

shinchiro commented:
^Yeah, almost same

wang-bin commented:
Look at this http://git.videolan.org/?p=vlc.git;a=commitdiff;h=e57241f6fa81855d34bb86e923792a0188cdbc10

The log says it fixes wrong colors on Sony Xperia XZ2. Maybe it's yuv. But I don't understand the change.

wm4 commented:
Status?

tmm1 commented:
This works and could be merged. However in recent Android releases there is a severe performance regression which makes this impractical to use on most devices.

https://issuetracker.google.com/issues/37079050

wm4 commented:
Ah, as usual, Google can do nothing right. Also, I can't even read your link, because it requires a login of some kind.
Wow, what an awful company.

wang-bin commented:
On android 8.0+ we can use AHardwareBuffer for both gles and vk, and java code(SurfaceTexture.onFrameAvailable) is no longer needed: ANativeWindow from AImageReader=>mediacodec=>AImageReader_ImageCallback=>AImageReader_acquireNextImageAsync=>AImage_getHardwareBuffer=>eglimage or vk resource. Haven't try it.

wm4 commented:
Looks like this PR won't go in then, at least not if Android doesn't handle the performance problem, or if wang-bin's suggested approach works.

sfan5 commented:
Due to a discussion in #ffmpeg-devel I looked into this again. This comment is just here to store my notes about understanding this stuff.

What this PR does:
* create a texture
* create SurfaceTexture + Surface java objects wrapping it
* give the surface for ffmpeg to let mediacodec render into
* when asked to map the texture
  * wait for frame to be available
  * retrieve stuff from the SurfaceTexture object
  * return the texture from step 1 as usual

The suggested approach with [AImageReader](https://developer.android.com/ndk/reference/group/media):
* also create (one?) texture
* create one with `AImageReader_newWithUsage` (usage=AIMAGE_FORMAT_PRIVATE)
* grab a window using `AImageReader_getWindow`
* configure mediacodec to render into it or with `AMediaCodec_setOutputSurface` [inside ffmpeg]
  * <b>since ff only uses the Java api which can't do this, how will this work?</b>
* call `AImageReader_acquireNextImageAsync` at some point(??)
  * or just use `AImageReader_acquireNextImage` (does that block?)
* when asked to map the texture
  * wait for the fence (= image available)
  * call `AImage_getHardwareBuffer`, keep AImage until we can free it next frame
  * call `eglCreateImageKHR` to create an egl image ([ref](https://www.khronos.org/registry/EGL/extensions/ANDROID/EGL_ANDROID_image_native_buffer.txt))
  * use `EGLImageTargetTexture2DOES` to bind the the image to the texture ([like here](https://github.com/mpv-player/mpv/blob/d7f6eba233369904ab1fa1ccd0f943bcf2fbc6e9/video/out/hwdec/hwdec_vaapi_gl.c#L168-L169))
  * return the texture from step 1 as usual

wnoun commented:
@sfan5  you may use `ANativeWindow_toSurface` to obtain the associated surface, or use the Java version ImageReader(use `AHardwareBuffer_fromHardwareBuffer` to get the corresponding `AHardwareBuffer`), and create `AVHWDeviceContext` like this PR.

wnoun commented:
for this pr, i got the correct output image, but sometimes it crashes, especially when terminating the player, looks like heap corruption(stacktrace shows that it happened in ta_free, may be called in ta desctuctor).

rcombs commented:
IMO any advancement to android hwdec should be done by adding a new `AV_PIX_FMT_AHARDWAREBUFFER` to lavc.

The biggest source of confusion here (imo) is that "Surface" and "Native Window" in Android really just mean "buffer queue". Here's an ffmpeg with a full NDK MediaCodec decode implementation, if anyone wants to poke around with it (the vendor-specific stuff is just to get a Binder pool running in cases where the process wasn't forked from zygote): https://downloads.plex.tv/ffmpeg-source/plex-media-server-ffmpeg-gpl-fa235d69f6.tar.gz

The biggest thing I'm unsure about in all this infrastructure is whether or not it works at all for frames > 8-bit depth.

sfan5 commented:
Other, existing APIs don't deal with >8-bit anyway so that's not a loss.
Testing has shown (https://github.com/mpv-android/mpv-android/issues/371#issuecomment-783419962) that 10-bit hwdec works with the current FFmpeg code _however_ the API only returns 8-bit frames anyway.
